src/core/provider/MCAccountValidator.cpp

```diff
@@ -0,0 +1,384 @@
+//
+//  MCAccountValidator.cpp
+//  mailcore2
+//
+//  Created by Christopher Hockley on 22/01/15.
+//  Copyright (c) 2015 MailCore. All rights reserved.
+//
+
+#include "MCAccountValidator.h"
+#include "MCMailProvider.h"
+#include "MCMailProvidersManager.h"
+#include "MCIMAPAsyncSession.h"
+#include "MCPOPAsyncSession.h"
+#include "MCSMTPAsyncSession.h"
+#include "MCNetService.h"
+#include "MCAddress.h"
+#include "MCIMAPOperation.h"
+#include "MCPOPOperation.h"
+#include "MCSMTPOperation.h"
+
+#include <arpa/inet.h>
+#include <resolv.h>
+
+using namespace mailcore;
+
+void AccountValidator::init()
+{
+    mEmail = NULL;
+    mUsername = NULL;
+    mPassword = NULL;
+    mOAuth2Token = NULL;
+    
+    mImapServices = new Array();
+    mSmtpServices = new Array();
+    mPopServices = new Array();
+    
+    mIdentifier = NULL;
+    mImapServer = NULL;
+    mPopServer = NULL;
+    mSmtpServer = NULL;
+    mImapError = ErrorNone;
+    mPopError = ErrorNone;
+    mSmtpError = ErrorNone;
+    
+    mCurrentServiceIndex = 0;
+    mCurrentServiceTested = 0;
+}
+
+AccountValidator::AccountValidator()
+{
+    init();
+}
+
+AccountValidator::~AccountValidator()
+{
+    MC_SAFE_RELEASE(mEmail);
+    MC_SAFE_RELEASE(mUsername);
+    MC_SAFE_RELEASE(mPassword);
+    MC_SAFE_RELEASE(mOAuth2Token);
+    MC_SAFE_RELEASE(mImapServices);
+    MC_SAFE_RELEASE(mSmtpServices);
+    MC_SAFE_RELEASE(mPopServices);
+    MC_SAFE_RELEASE(mIdentifier);
+}
+
+void AccountValidator::start()
+{
+    if (setup()) test();
+    else mCallback->operationFinished(this);
+}
+
+void AccountValidator::operationFinished(Operation * op)
+{
+    opCompleted();
+}
+
+bool AccountValidator::setup()
+{
+    if (mEmail == NULL) {
+        if (mUsername == NULL) return false;
+        else mEmail = mUsername;
+    }else if (mUsername == NULL){
+        mUsername = mEmail;
+    }
+    
+    MailProvider *provider = MailProvidersManager::sharedManager()->providerForEmail(mUsername);
+    
+    if (provider == NULL) {
+        Array * components;
+        String * domain;
+    
+        components = mUsername->componentsSeparatedByString(MCSTR("@"));
+        if (components->count() < 2)
+            return false;
+    
+        domain = (String *) components->lastObject();
+        provider = VTMxRecordForHostname(domain);
```

Comment:
VTMxRecordForHostname() should probably not run in the main thread because it's performing a network call. I'd suggest creating an operation and adding an OperationQueue for that.